skimage/filters/thresholding.py

```diff
@@ -1132,49 +1142,38 @@ def threshold_multiotsu(image, classes=3, nbins=256):
+        msg = ("threshold_multiotsu is expected to work correctly only for "
+               "grayscale images; image shape {0} looks like an RGB image")
+        warn(msg.format(image.shape))
+
     # calculating the histogram and the probability of each gray level.
-    hist, bin_centers = histogram(image.ravel(),
+    prob, bin_centers = histogram(image.ravel(),
                                   nbins=nbins,
-                                  source_range='image')
-    prob = hist / image.size
-    # histogram ignores nbins for integer arrays.
-    nbins = len(bin_centers)
-
-    # Compute the zeroth (momP, cumulative probability) and first
-    # (momS, mean) moments
-
-    # building the lookup tables.
-    # step 1: calculating the diagonal.
-    prob[0] = 0
-    momP = np.diagflat(prob)
-    momS = np.diagflat(np.arange(nbins) * prob)
-
-    # step 2: calculating the first row.
-    momP[1, 2:] = prob[1] + np.cumsum(prob[2:])
-    momS[1, 2:] = prob[1] + np.cumsum(np.arange(2, nbins) * prob[2:])
-
-    # step 3: the other rows are recursively computed as:
-    # A[i, j] = A[1, j] - A[1, i-1] for A in {momP, momS};  i > 1 and j > i.
-    upper_tri = np.triu_indices_from(momP[2:, 2:], 1)
-
-    momP[2:, 2:][upper_tri] = (momP[1, 2:][upper_tri[1]]
-                               - np.repeat(momP[1, 1:-2],
-                                           np.arange(nbins - 3, 0, -1)))
-    momS[2:, 2:][upper_tri] = (momS[1, 2:][upper_tri[1]]
-                               - np.repeat(momS[1, 1:-2],
-                                           np.arange(nbins - 3, 0, -1)))
-
-    # step 4: calculating the between class variance.
-    var_btwcls = np.zeros_like(momP)
-    idx = momP > 0
-    var_btwcls[idx] = momS[idx] ** 2 / momP[idx]
-
-    # finding max threshold candidates, depending on classes.
-    # number of thresholds is equal to number of classes - 1.
-    aux_thresh = _find_threshold_multiotsu(var_btwcls, classes, nbins)
-
-    # correcting threshold values.
-    idx_thresh = bin_centers[:-1][aux_thresh.astype('int')]
-
-    return idx_thresh
+                                  source_range='image',
+                                  normalize=True)
+    prob = prob.astype('float32')
+
+    nvalues = np.count_nonzero(prob)
+    if nvalues < classes:
+        msg = ("The input image has only {} different values. "
+               "It can not be thresholded in {} classes")
+        raise ValueError(msg.format(nvalues, classes))
+    elif nvalues == classes:
+        thresh_idx = np.where(prob > 0)[0][:-1]
+    else:
+        # Get threshold indices
+        try:
+            thresh_idx = _get_multiotsu_thresh_indices_lut(prob, classes - 1)
+        except MemoryError:
+            # Don't use LUT if the number of bins is too large (if the
+            # image is uint16 for example): in this case, the
+            # allocated memory is too large.
+            thresh_idx = _get_multiotsu_thresh_indices(prob, classes - 1)
```

Comment:
I'm unclear about this part. What are the similarities and differences between _get_multiotsu_thresh_indices_lut and _get_multiotsu_thresh_indices. The docstrings are not very informative.

Do we expect the same results to be returned? Is *lut just faster with more memory consumption? Then, can we add tests to check that returned values are identical?

Also, is the try except approach suitable here? It means that using this function on multiple large images will fill up the memory, fail, unload the memory, perform the other calculation... It's not so efficient. Based on the answers of the question I pointed out, I think we can find another approach.

Thank you!

Reply:
Hi @sciunto,
yes, the result is supposed to be the same. The use of LUT is in fact faster at the price of more memory consumption. This can be added to the doc.
I can in fact also assert in a test that the returned values are identical with both approaches.

The memory will not be filled up: the necessary memory is allocated in `_get_multiotsu_thresh_indices_lut` (for `var_btwcls`, `zeroth_moment` and `first_moment`), the OS is asked for memory allocation, and if there is not enough, the `MemoryError` is raised.

The amount of required memory doesn't depend on the size of the input image but on the number of bins `nbins`. This number is automatically set by the `histogram` function when the input image values are integrals. That's why we face problems with `uint16` images (see #3975).

The amount of available RAM is machine dependent, that's why I think that the `try except` is the best approach.

Reply:
Thanks for the explanations! Then :+1: for adding this to the code :)